### PR TITLE
Propose normalised structure for new cell group markers

### PR DESCRIPTION
**Pull request by pinin4fjords**
This is a starting point for discussion on incorporating markers for generic groups of cells, to include cluster, cell type and other cell group markers (for example for later differential expression efforts). 

Since we're no longer dealing with integer values for clusterings and clusters, I though perhaps now would be a good time to normalise to save excess storage use for VARCHAR fields. I've proposed a table to define cell groupings for an experiment (clusterings and other), with an integer auto-incremented primary key to be referenced as a foreign key in both cell cluster assignments and cell group marker storage. 

This is derived from my possibly out-of-date early-career experience in DB design, and that was MySQL rather than Postgres, so it's possible this is not the way to go- but it's a start.

### CODE REVIEW

File: db/gxasc-schema.sql [hunk 1]
```diff
@@ -42,6 +42,33 @@ CREATE TABLE scxa_cell_clusters
   PRIMARY KEY (experiment_accession, k, cell_id)
 );
 
+CREATE TABLE scxa_cell_group
```
Comment (pcm32) [inline db/gxasc-schema.sql:45]: All these changes should go on a new file on the migrations part, not in the baseline schema.

Comment (pinin4fjords) [reply db/gxasc-schema.sql:45]: Yep, I was more using this to illustrate the structure I wanted to go for, for discussion, don't actually want to merge this.

Comment (pcm32) [reply db/gxasc-schema.sql:45]: I'm overall fine with the structure, but we need to see how it constraints both the loading and querying (and whether for instance the required JOIN on query time is worth the space saving due to normalisation). We could make a ball park estimate on the size of the denormalised table vs normalised for the largest experiments.

Comment (pinin4fjords) [reply db/gxasc-schema.sql:45]: @alfonsomunozpomer - do you have any feelings on this? I was always taught to normalise in this sort of situation, but I have no feeling for the performance implications here.

Comment (pcm32) [reply db/gxasc-schema.sql:45]: I would also suggest that we open a parallel PR on db-scxa where you start sketching how the loading will work, including testing some simple examples, which in turn will provide things to keep in mind for the structure being proposed here.

Comment (pcm32) [reply db/gxasc-schema.sql:45]: our primary objective should be best perfomance at query time, followed by best perfomance at loading time, followed by size considerations and then db structure correctness (IMHO). When any of this is massively bad, it should take slight precedence over the others until fixed I would say. And query time should never be bad at all. Currently loading time for clustering data is slow in the current setup for large experiments.

Comment (pinin4fjords) [reply db/gxasc-schema.sql:45]: > I would also suggest that we open a parallel PR on db-scxa where you start sketching how the loading will work, including testing some simple examples, which in turn will provide things to keep in mind for the structure being proposed here.

This ongoing at https://github.com/ebi-gene-expression-group/db-scxa/pull/40. I've made proposals for loading of clusterings, markers next.

File: flyway/scxa/migrations/V9__scxa-allow-generic-markers.sql [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+  cell_group_id integer references scxa_cell_group(id)
+  marker_probability DOUBLE PRECISION,
+  CONSTRAINT scxa_cell_group_marker_genes_pkey,
+  PRIMARY KEY (gene_id, cell_group_id),
+  UNIQUE (gene_id, cell_group_id)  
+);
+
+-- Create a table from data prepared during bundling, replacing the previously used materialised view
+
+CREATE TABLE scxa_cell_group_marker_gene_stats
```
Comment (pcm32) [inline flyway/scxa/migrations/V9__scxa-allow-generic-markers.sql:33]: @pinin4fjords talk to Alfonso about the use of the two materialised views (stats and not stats), as my impression is that one of them is used to construct the second, but then they only query the second one, so we might not need both in the final scheme (but just the one that they query).

File: flyway/scxa/migrations/V9__scxa-allow-generic-markers.sql [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+  cell_group_id integer references scxa_cell_group(id),
+  marker_probability DOUBLE PRECISION,
+  UNIQUE (gene_id, cell_group_id)  
+);
+
+-- Create a table from data prepared during bundling, replacing the previously used materialised view
+
+CREATE TABLE scxa_cell_group_marker_gene_stats
+(
+  gene_id VARCHAR(255) NOT NULL,
```
Comment (pcm32) [inline flyway/scxa/migrations/V9__scxa-allow-generic-markers.sql:33]: What about the expression of the genes in the other groups where they are not marker genes? Remember that in the heatmap, for gene A that is a marker gene in group Ga, you will also need summarised expression in groups Gb, Gc, ..., Gn. Conversely, for gene B, which is a marker in Gb, you will also need summarised expression in Ga... Gn.

Comment (pcm32) [reply flyway/scxa/migrations/V9__scxa-allow-generic-markers.sql:33]: ahhh, ok, I see how this is covered.

File: flyway/scxa/migrations/V9__scxa-allow-generic-markers.sql [hunk 4]
```diff
@@ -0,0 +1,40 @@
+
+CREATE TABLE scxa_cell_group
+(
+  id SERIAL not null PRIMARY KEY,
+  experiment_accession VARCHAR(255) NOT NULL,
+  variable VARCHAR(255) NOT NULL,
+  value VARCHAR(255) NOT NULL,
+  UNIQUE (experiment_accession, variable, value)  
+);
```
[the remark this answers is not in the document]
Comment (pinin4fjords) [reply flyway/scxa/migrations/V9__scxa-allow-generic-markers.sql:9]: Ahh, I should set all the FKs to cascade actually. 

But what do you mean for this specific case, given that there is not a foreign key to the experiment table?

[the remark this answers is not in the document]
Comment (pinin4fjords) [reply flyway/scxa/migrations/V9__scxa-allow-generic-markers.sql:9]: I could make it an FK I guess- will just do that and test...

**Pull request merged**